### PR TITLE
Add incident details option for json and sarif formats

### PR DESCRIPTION
**Pull request by fnareoh**
## Context

- Add `--with-incident-details` to output (json and sarif output only) more information about incident when the secret is already known by the gitguardian dashboard.

## PR check list

- [ ] As much as possible, the changes include tests (unit and/or functional)
- [ ] If the changes affect the end user (new feature, behavior change, bug fix) then the PR has a changelog entry (see doc/dev/getting-started.md). If the changes do not affect the end user, then the `skip-changelog` label has been added to the PR.

### CODE REVIEW

File: tests/unit/verticals/secret/output/test_json_output.py [hunk 1]
```diff
@@ -495,3 +498,89 @@ def test_ignore_known_secrets(verbose, ignore_known_secrets, secrets_types):
...
+            filemode=Filemode.NEW,
+        ),
+        scan=deepcopy(TWO_POLICY_BREAKS),  # 2 policy breaks
+    )
+
+    all_policy_breaks = result.scan.policy_breaks
+
+    known_policy_breaks = []
+
+    # add known_secret for the secrets that are known, when the option is, the known_secret field is not returned
```
Comment (agateau-gg) [inline tests/unit/verticals/secret/output/test_json_output.py:540]: I don't understand this comment, maybe a word is missing?

File: tests/unit/verticals/secret/output/test_json_output.py [hunk 2]
```diff
@@ -495,3 +498,89 @@ def test_ignore_known_secrets(verbose, ignore_known_secrets, secrets_types):
         assert not incident_for_policy_break_type[policy_break.break_type][
             "incident_url"
         ]
+
+
+@pytest.mark.parametrize("verbose", [True, False])
```
Comment (agateau-gg) [inline tests/unit/verticals/secret/output/test_json_output.py:503]: If I am not mistaken, the behavior does not depend on whether verbose is enabled or not, maybe we can remove this?

File: changelog.d/20240906_141336_fnareoh_scrt_4693_ggshield_output_expose_incident_date_and_severity.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+### Added
+
+- Add `--with-incident-details` to output (json and sarif output only) more information about incident when the secret is already known by the gitguardian dashboard.
```
Comment (agateau-gg) [inline changelog.d/20240906_141336_fnareoh_scrt_4693_ggshield_output_expose_incident_date_and_severity.md:3]:
```suggestion
- Add `--with-incident-details` to output more information about known incidents (JSON and SARIF outputs only).
```

File: ggshield/cmd/secret/scan/secret_scan_common_options.py [hunk 4]
```diff
@@ -120,6 +120,13 @@ def _banlist_detectors_callback(
     metavar="DETECTOR",
 )
 
+_with_incident_details_option = click.option(
+    "--with-incident-details",
+    is_flag=True,
+    help="Display full details about the dashboard incident if one is found (json and sarif format only).",
```
Comment (agateau-gg) [inline ggshield/cmd/secret/scan/secret_scan_common_options.py:126]:
```suggestion
    help="Display full details about the dashboard incident if one is found (JSON and SARIF formats only).",
```

File: ggshield/verticals/secret/output/secret_sarif_output_handler.py [hunk 5]
```diff
@@ -93,6 +103,10 @@ def _create_sarif_result_dict(
     }
     if policy_break.incident_url:
         dct["hostedViewerUri"] = policy_break.incident_url
+        if policy_break.incident_url in incident_details:
+            dct["incidentDetails"] = incident_details[
```
[the remark this answers is not in the document]
Comment (fnareoh) [reply ggshield/verticals/secret/output/secret_sarif_output_handler.py:107]: Sorry I should have looked into that before. I get the impression that for any additional information, the key `properties` is used as a catch all [(github did this for their tag and precision additional fields)](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning), what do you think of having the incident details nested in properties ?

[the remark this answers is not in the document]
Comment (agateau-gg) [reply ggshield/verticals/secret/output/secret_sarif_output_handler.py:107]: Yes [property bags](https://github.com/microsoft/sarif-tutorials/blob/main/docs/2-Basics.md#property-bags) would be appropriate.

File: ggshield/verticals/secret/output/secret_json_output_handler.py [hunk 6]
```diff
@@ -108,6 +123,10 @@ def serialized_policy_break(
         if policy_breaks[0].known_secret:
             flattened_dict["known_secret"] = policy_breaks[0].known_secret
             flattened_dict["incident_url"] = policy_breaks[0].incident_url
+            if policy_breaks[0].incident_url in incident_details:
+                flattened_dict["incident_details"] = incident_details[
+                    policy_breaks[0].incident_url
+                ]
```
[the remark this answers is not in the document]
Comment (fnareoh) [reply ggshield/verticals/secret/output/secret_json_output_handler.py]: Unfortunately this does not work due to [the Base class in pygitguardian having a weird __bool__ definition](https://github.com/GitGuardian/py-gitguardian/blob/215f5294a8efd4cb7e0c89472f00d955587b341e/pygitguardian/models.py#L81) :'(

File: ggshield/verticals/secret/secret_scan_collection.py [hunk 7]
```diff
@@ -177,3 +178,17 @@ def get_all_results(self) -> Iterable[Result]:
...
+        incidents_url_with_details: dict[str, SecretIncident] = {}
+        for result in self.get_all_results():
+            for policy_break in result.scan.policy_breaks:
+                url = policy_break.incident_url
+                if url and url not in incidents_url_with_details:
+                    incident_id = int(url.split("/")[-1])
+                    resp = client.retrieve_secret_incident(
+                        incident_id, with_occurrences=0
+                    )
+                    if type(resp) == SecretIncident:
```
[the remark this answers is not in the document]
Comment (agateau-gg) [reply ggshield/verticals/secret/secret_scan_collection.py:193]: That's a good idea, but maybe you can put it in `core.errors`?

**Review by agateau-gg — CHANGES_REQUESTED** [review]

File: ggshield/verticals/secret/secret_scan_collection.py [hunk 8]
```diff
@@ -177,3 +178,17 @@ def get_all_results(self) -> Iterable[Result]:
             for scan in self.scans:
                 if scan.results:
                     yield from scan.results.results
+
+    def get_incident_details(self, client: GGClient) -> Dict[str, SecretIncident]:
+        incidents_url_with_details: dict[str, SecretIncident] = {}
```
[the remark this answers is not in the document]
Comment (fnareoh) [reply ggshield/verticals/secret/secret_scan_collection.py:184]: Arf sorry !

File: ggshield/verticals/secret/output/secret_json_output_handler.py [hunk 9]
```diff
@@ -65,6 +51,19 @@ def create_scan_dict(
                 scan_dict["total_occurrences"] += inner_scan_dict["total_occurrences"]
         return scan_dict
 
+    def _process_scan_impl(self, scan: SecretScanCollection) -> str:
+        incident_details = (
+            scan.incident_details(self.client)
+            if self.with_incident_details and self.client
```
[the remark this answers is not in the document]
Comment (fnareoh) [reply ggshield/verticals/secret/output/secret_json_output_handler.py]: I added an assert before but pyright does not really get it so I had to leave the "and self.client"

**Review by agateau-gg — CHANGES_REQUESTED** [review]

**Review by agateau-gg — APPROVED** [review]

**Pull request merged**